src/teleop/gui/src/components/ROSEcho.vue

```diff
@@ -0,0 +1,230 @@
+  <script>
+  
+  import ROSLIB from "roslib"
+
+  export default {
+    name: 'ROSEcho',
+    mounted() {
+        this.populateTopics()
+    },
+    data() {
+        return {
+            topics: [],
+            selectedTopics : [],
+            feed: [],
+            cols: [],
+            topicsService: null,
+            serviceRequest: null
+        }
+    },
+    
+
+    methods: {
+        populateTopics : function(){ //populates active topics to the display
+            this.serviceClient.callService(this.serviceRequest, (result) => {
+                for(var i = 0; i < result.topics.length; i++){
+                    this.topics.push(new ROSLIB.Topic({
+                    ros: this.$ros,
+                    name: result.topics[i],
+                    messageType: result.types[i]
+                    }));
+                }
+            });
+
+        },
+
+        addType : function(){ //method to handle checking/unchecking the topics
+            //Add newly added topics to the column display and subscribe
+            this.selectedTopics.forEach(prev => {
+                const found = this.cols.find(newTopic => newTopic.name === prev.name);
+                if(!found){
+                    prev.subscribe((msg) => {
+                        var topicCol = this.cols.find(c => c.name === prev.name);
+                        if(topicCol) topicCol.messages.push(JSON.stringify(msg, null, '\t'));
+                        if(topicCol.messages.length > 3) topicCol.messages.splice(0,1);
+                    });
+                    this.cols.push({name: prev.name, messages: [], muted: false});
+                }
+            });
+            
+            //Remove topics from column display if unchecked and unsubscribe
+            this.cols.forEach(prev => {
+                const found = this.selectedTopics.find(newTopic => newTopic.name === prev.name);
+                if(!found){
+                    var index = this.cols.findIndex(obj => obj === prev);
+                    this.selectedTopics[index].unsubscribe();
+                    this.cols.splice(index, 1);
+                }
+            });
+        },
+
+        mute: function(c) {
+            c.muted = !c.muted;
+        }
+
+    },
+  
+  created: function () {
+
+    this.serviceClient = new ROSLIB.Service({
+        ros : this.$ros,
+        name : '/rosapi/topics',
+        serviceType : 'rosapi/Topics'
+    });
+
+    this.serviceRequest = new ROSLIB.ServiceRequest();
+
+    let interval = window.setInterval(() => { //auto-refreshes so newly activated topics (custom topics) are displayed
+            this.serviceClient.callService(this.serviceRequest, (result) => {
+                if(this.topics.length != result.topics.length){
+                    this.topics.push(new ROSLIB.Topic({
+                        ros: this.$ros,
+                        name: result.topics[result.topics.length-1],
+                        messageType: result.types[result.topics.length-1]
```

Comment:
This looks to be operating under the assumption that we never see more than one new topic at a time, and also that it would be added to the end of result.topics instead of in the middle. These assumptions make me a little nervous, would it be terrible to just call this.populateTopics()?

Reply:
Yeah I agree with this, would just have to add a this.topics = [] to the beginning of this.populateTopics()